SwiftCompilerSources/Sources/AST/Declarations.swift

```diff
@@ -165,3 +267,137 @@ extension Optional where Wrapped == Decl {
     OptionalBridgedDeclObj(self?.bridged.obj)
   }
 }
+
+public typealias AccessLevel = swift.AccessLevel
+
+public typealias Identifier = swift.Identifier
+
+public typealias GenericTypeParamKind = swift.GenericTypeParamKind
+
+public typealias ASTContext = BridgedASTContext
+
+public typealias Expr = BridgedExpr
+
+public typealias SourceFile = BridgedSourceFile
+
+public typealias FileUnit = BridgedFileUnit
+
+public class GenericParamList {
+  public var bridged: BridgedGenericParamList
+  public init(bridged: BridgedGenericParamList) { self.bridged = bridged }
+}
+
+public typealias TrailingWhereClause = BridgedTrailingWhereClause
+
+public class ParameterList : RandomAccessCollection {
+  public class Iterator : IteratorProtocol {
+    public typealias Element = ParamDecl
+    private var index : Int = 0
+    private let parameterList : ParameterList
+    init(parameterList : ParameterList) { self.parameterList = parameterList }
+    public func next() -> Element? {
+      if index < parameterList.bridged.size {
+        return parameterList.bridged.get(index).asDecl.declObj.getAs(ParamDecl.self)
+      }
+      return nil
+    }
+  }
+  public typealias Index = Int
+  public var startIndex: Index { 0 }
+  public var endIndex: Index { bridged.size }
+  public func index(after i: Index) -> Index { i + 1 }
+
+  public func makeIterator() -> Iterator {
+    Self.Iterator(parameterList: self)
+  }
+
+  public var bridged: BridgedParameterList
+
+  public init(bridged: BridgedParameterList) {
+    self.bridged = bridged
+  }
+
+  public subscript(_ index: Index) -> ParamDecl {
+    return bridged.get(index).asDecl.declObj.getAs(ParamDecl.self)
+  }
+
+  public static func create(
+    _ astContext: ASTContext, leftParenLoc: SourceLoc?, parameters: [ParamDecl],
+    rightParenLoc: SourceLoc?
+  ) -> ParameterList {
+    ParameterList(bridged: parameters.map{BridgedParamDecl(raw: $0.bridged.obj)}.withBridgedArrayRef {
+      BridgedParameterList.createParsed(
+        astContext, leftParenLoc: leftParenLoc.bridgedLocation, parameters: $0,
+        rightParenLoc: rightParenLoc.bridgedLocation)
+    })
+  }
+}
+
+extension GenericParamList {
+  public static func create(
+    _ astContext: ASTContext, leftAngleLoc: SourceLoc?, parameters: [GenericTypeParamDecl],
+    genericWhereClause: TrailingWhereClause?,
+    rightAngleLoc: SourceLoc?
+  ) -> GenericParamList {
+    let paramsNew = parameters.map{ ASTBridging.BridgedGenericTypeParamDecl(raw: $0.bridged.obj) }
+    return paramsNew.withBridgedArrayRef {
+      GenericParamList(bridged: BridgedGenericParamList.createParsed(
+        astContext, leftAngleLoc: leftAngleLoc.bridgedLocation, parameters: $0,
+        genericWhereClause: genericWhereClause.bridged, rightAngleLoc: rightAngleLoc.bridgedLocation
+      ))
+    }
+  }
+}
+
+extension BridgedDecl {
+  public var declObj: BridgedDeclObj {
+    BridgedDeclObj(SwiftObject(raw.bindMemory(to: BridgedSwiftObject.self, capacity: 1)))
```

Comment:
Maybe better to do this ugly type casting on the C++ side by adding a constructor `BridgedDeclObj(BridgedDeclObj)`